sentry-ruby/lib/sentry/transport.rb

```diff
@@ -83,27 +92,85 @@ def generate_auth_header
 
     def encode(event)
       # Convert to hash
-      event_hash = event.to_hash
+      event_payload = event.to_hash
 
-      event_id = event_hash[:event_id] || event_hash["event_id"]
-      item_type = get_item_type(event_hash)
+      event_id = event_payload[:event_id] || event_payload["event_id"]
+      item_type = get_item_type(event_payload)
+
+      envelope_header = {
+        event_id: event_id,
+        dsn: @dsn.to_s,
+        sdk: Sentry.sdk_meta,
+        sent_at: Sentry.utc_now.iso8601
+      }
+
+      event_header = { type: item_type, content_type: 'application/json' }
 
       envelope = <<~ENVELOPE
-        {"event_id":"#{event_id}","dsn":"#{@dsn.to_s}","sdk":#{Sentry.sdk_meta.to_json},"sent_at":"#{Sentry.utc_now.iso8601}"}
-        {"type":"#{item_type}","content_type":"application/json"}
-        #{JSON.generate(event_hash)}
+        #{JSON.generate(envelope_header)}
+        #{JSON.generate(event_header)}
+        #{JSON.generate(event_payload)}
+        #{fetch_pending_client_report}
       ENVELOPE
 
       log_info("Sending envelope [#{item_type}] #{event_id} to Sentry")
 
       envelope
     end
 
+    # valid reasons are
+    #   :ratelimit_backoff
+    #   :queue_overflow
+    #   :cache_overflow - NA
+    #   :network_error
+    #   :sample_rate
+    #   :before_send
+    #   :event_processor
```

Comment:
Should we link to the develop docs?

Also what do you think about enforcing an allow list here for valid reasons?